### DISCUSSION TITLE
v0.11.9 Testing checklist

### DISCUSSION THREAD

**Discussion by jorgeepc** [kubeshop/tracetest#2557] [Release testing]
# Testing Tracetest

Latest release: https://github.com/kubeshop/tracetest/releases/tag/v0.11.9-rc.3

This is the entire checklist on what we should do to assert that Tracetest is working fine on each version release. On each version release, we can copy the contents of this template checklist and open a Github Discussion to start each test.

On Tracetest, we work in two ways to test the entire system and guarantee that everything is working fine. One is through automatic tests, where we run unit, integration and end-to-end tests on Tracetest CLI, Web UI and Server.

Another source of tests is the manual tests that we execute on each release, simulating users, and checking if everything is ok on these tests.

## Manual Tests

On our manual tests, we aim to do some [sanity checks](https://en.wikipedia.org/wiki/Sanity_check) to see if the main features are working as expected. Usually, we run it on each Tracetest release.

### Tests to validate on Release Candidate

- Test server installation via CLI
  - [x] Docker Compose and no demo API (@mathnogueira on Linux)
  - [x] Docker Compose and demo API (@mathnogueira on Linux)
  - [x] Kubernetes and no demo API (@schoren)
  - [x] Kubernetes and demo API (@schoren)
  - [x] Verify that Tracetest is not sending analytics if you have the `TRACETEST_DEV` env for docker install (@mathnogueira)
  - [x] Verify that Tracetest is not sending analytics if you have the `TRACETEST_DEV` env for k8s install  (@schoren)

- Test DataStore provisioning and basic test
  - [x] Open WebUI and go to `/settings` page. The provisioned Data Store should be selected
  - [x] Create a test on WebUI that calls a demo API (like [Pokeshop](https://docs.tracetest.io/live-examples/pokeshop/overview) or [Open Telemetry Store](https://docs.tracetest.io/live-examples/opentelemetry-store/overview)). This test should fetch traces correctly and run without errors

- Test Tracetest examples (**_We haven't changed code related to this, prob we can skip this step_**)
  - [x] [Jaeger example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-jaeger)
  - [x] [Tempo example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-tempo)
  - [x] [New Relic example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-new-relic)
  - [x] [Lightstep example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-lightstep)
  - [x] [OpenSearch example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-opensearch)
  - [x] [Elastic APM example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-elasticapm-with-elastic-agent)
      - Tracetest is working, however, we found a bug due to an update on Elastic infra: https://github.com/kubeshop/tracetest/issues/2591
  - [x] [SignalFX example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-signalfx)
  - [x] [Datadog example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-datadog)
  - [ ] [Amazon X-Ray example](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-amazon-x-ray)
  - [x] [Honeycomb](https://github.com/kubeshop/tracetest/tree/main/examples/tracetest-honeycomb)

- Test specific features added/changed on this release:
  - [x] Validate fix for `apply environment --file`. We now should be able to create a new env with id set on file, and also update it.  (@schoren)
  - [x] Validate new CLI `transaction` resource: `list transaction`, `apply transaction`, `get transaction`, `delete transaction`  (@schoren)
  - [x] Transactions CLI command: `tracetest apply transaction -f [new-transaction-file]`
    - has error with relative paths. fix: https://github.com/kubeshop/tracetest/pull/2582
  - [x] Transactions CLI command: `tracetest apply transaction -f [existing-transaction-file]`  (@schoren)
  - [x] Transactions CLI command: `tracetest delete transaction --id [existing-id]`  (@schoren)
  - [x] Transactions CLI command: `tracetest delete transaction --id [non-existing-id]`  (@schoren)
  - [x] Transactions CLI command: `tracetest export transaction --id current --file [transaction-file]`
    - **Not working**: https://github.com/kubeshop/tracetest/issues/2586
  - [x] Transactions CLI command: `tracetest get transaction --id [non-existing-id]`  (@schoren)
  - [x] Transactions CLI command: `tracetest get transaction --id [existing-id] --output pretty` (@schoren)
  - [x] Transactions CLI command: `tracetest get transaction --id [existing-id] --output json`  (@schoren)
  - [x] Transactions CLI command: `tracetest get transaction --id [existing-id] --output yaml`  (@schoren)
  - [x] Transactions CLI command: `tracetest list transaction --output pretty`  (@schoren)
  - [x] Transactions CLI command: `tracetest list transaction --output json`  (@schoren)
  - [x] Transactions CLI command: `tracetest list transaction --output yaml`  (@schoren)
  - [x] Transactions CLI command: `tracetest list transaction --skip 1 --take 2`  (@schoren)
  - [x] Transactions CLI command: `tracetest list transaction --sortBy name --sortDirection desc`  (@schoren)
  - [x] Validate CLI `test run` command using a transaction definition  (@schoren)
  - [x] Validate webUI transaction CRUD (@schoren)
  - [x] Validate webUI transaction run (@schoren)
  - [x] Validate output from CLI `list`, `apply`, `get`, `delete` for resource `datastore`. Related to [output normalization](https://github.com/kubeshop/tracetest/pull/2475) (@schoren, @danielbdias )
    - panic on server https://github.com/kubeshop/tracetest/pull/2584
    - error when listing in no-tracing mode https://github.com/kubeshop/tracetest/pull/2585
  - [x] Validate output from CLI `list`, `apply`, `get`, `delete` for resource `environment`. Related to [output normalization](https://github.com/kubeshop/tracetest/pull/2475) (@schoren)
    - panic on server https://github.com/kubeshop/tracetest/pull/2584
  - [x] Validate output from CLI `list`, `apply`, `get`, `delete` for resource `transaction`. Related to [output normalization](https://github.com/kubeshop/tracetest/pull/2475) (@schoren)
  - [x] Validate CLI [dashboard command](https://github.com/kubeshop/tracetest/pull/2490) (@schoren)
  - [x]  Validate CLI [version command](https://github.com/kubeshop/tracetest/pull/2505) (@schoren)
  - [x] Validate [context flag](https://github.com/kubeshop/tracetest/pull/2477) on `server install` command using `--run-environment kubernetes --mode with-demo --kubernetes-context xxxx` (@schoren)
  - [x] Validate [context flag](https://github.com/kubeshop/tracetest/pull/2477) on `server install` command using `--run-environment kubernetes --mode just-tracetest --kubernetes-context xxxx`
  - [x] Validate [context flag](https://github.com/kubeshop/tracetest/pull/2477) on `server install` command using `--run-environment docker --mode with-demo`
  - [x] Validate [context flag](https://github.com/kubeshop/tracetest/pull/2477) on `server install` command using `--run-environment docker --mode just-tracetest`
  - [x] Validate if it's possible to have attributes with dashes in it's name
  
### Tests to validate on Final Release

- Pipeline
  - [x] Check if our release pipeline on [Release Tracetest](https://github.com/kubeshop/tracetest/actions/workflows/release-version.yml) workflow on Github Actions worked correctly

- Test CLI updates
  - [x] MacOS via homebrew
  - [x] MacOS via curl script
  - [x] Windows via chocolatey
  - [x] Windows via manual download

- Docs
  - [x] Double check [Detailed installation](https://docs.tracetest.io/getting-started/detailed-installation) docs and see if everything is documented correctly

- Test specific features added/changed on this release:
  - [ ] Add list of features here
  
## Automatic Tests

Today Tracetest has 3 main components: a WebUI, a CLI and a Server.

### Web UI

- **Unit tests**: Run by executing `npm test` on `./web` folder
- **End-to-end tests**: Run using [cypress](https://www.cypress.io/) against a temporary Tracetest created on Kubernetes. 

### CLI

- **Unit tests**: Run by executing `make test` on `./cli` folder

### Server

- **Unit tests**: Run by executing `make test` on `./server` folder
- **Integration tests**: Run along with some unit tests running `make test` on `./server` folder, but also done in a matrix test on Github actions, by executing the `test-examples` action.
- **End-to-end tests**: Run using Tracetest to test itself. We deploy two instances of Tracetest and use one to test API calls to another on Github actions, by executing the `trace-testing` action.

**Comment by danielbdias** [comment]
`list datastore` with empty datastores return error: 
<img width="733" alt="image" src="https://github.com/kubeshop/tracetest/assets/12397879/e55518b4-c868-4432-9cca-cffb6eb61559">

**Comment by schoren** [comment]
panic on server when getting non-augmentable entities as json
![Screenshot 2023-05-24 at 19 16 42](https://github.com/kubeshop/tracetest/assets/314548/cd392b3a-da6b-4ca6-b5fb-ae24b97a7244)

**Comment by schoren** [comment]
bug with exporting transactions: https://github.com/kubeshop/tracetest/issues/2586
